readme.md
--------

## AUTOMATIC ATTACHMENT ROTATION (AND RESIZE) v5.7

[**By Dougiefresh**](http://www.simplemachines.org/community/index.php?action=profile;u=253913) -> [Link to Mod](http://custom.simplemachines.org/mods/index.php?mod=4087)

--------

## Introduction
This mod allows the automatic rotation and/or flipping of images **ONLY IF** the EXIF information contained within (if such exists) indicates that such processing is required in order to show the picture  in the correct orientation.

**NOTE:** SMF contains an option called **Re-encode potentially dangerous image attachments**, which is turned on by default.  This re-encoding removes the orientation information from attachments (amongst other things), which means attachments uploaded prior to this mod being installed will **NOT** be able to rotated properly, as the EXIF information is missing from the re-encoded attachment file!

### Major changes introduced in v4.0

- Option to enable/disable automatic image rotation.
- Code update to enable automatic image resizing/reformatting in posts (and PMs if the [PM Attachments](https://custom.simplemachines.org/mods/index.php?mod=1974) mod is installed).

### Major change introduced in v5.0

- Option to automatically, manually or batch resize/reformat existing images using 'Attachment Settings' options (non-JPEG images will only be reformatted to JPEG if the *Reformat non-JPEG images to JPEG* option is enabled).

## Post Screen Changes
Beside each image attachment a dropbox will be displayed listing the following rotation options:

- No Change
- Rotate 90[sup]o[/sup] Right
- Rotate 90[sup]o[/sup] Left
- Rotate 180[sup]o[/sup]
- Horizontal Flip
- Vertical Flip
- Horizontal Flip, Rotate 90[sup]o[/sup] Right
- Vertical Flip, Rotate 90[sup]o[/sup] Right

## Automatic Resizing of Existing Images
When a post is dispayed, and if the Attachment Setting *Resize existing images* option is enabled, images in the post will be automatically resized (and reformatted to JPEG if the Attachment Setting option *Reformat non-JPEG images to JPEG* is enabled) using the same Attachment Setting options (ie, Reformat non-JPEG images to JPEG, JPEG quality factor and/or maximum width/height) that are applied to new image attachments in posts.  In addition, if the Attachment Setting option *Create backup of original image file when resizing* is also enabled the original image file will be saved to the attachments directiory with the extension '.rei'.

## Manual Resizing of Existing Images
This feature, which can be accessed via **Admin => Forum => Attachments and Avatars => Browse Files => Resize Existing Images**, can be used to selectively resize/reformat existing images using the same Attachment Setting options (ie, Reformat non-JPEG images to JPEG, JPEG quality factor and/or maximum width/height) that are applied to new image attachments in posts.

## Batch Resizing of Existing Images
This feature, which can be accessed via **Admin => Forum => Attachments and Avatars => File Maintenance => Batch Resize Existing Images**, will resize/reformat **all** existing images using the same Attachment Setting options (ie, Reformat non-JPEG images to JPEG, JPEG quality factor and/or maximum width/height) that are applied to new image attachments in posts.
 
The forum should be configured as follows **prior** to commencing batch resizing:
o *Admin => Configuration => Server Settings => General => Enable Maintenance Mode* should be **enabled**.
o *Admin => Configuration => Security and Moderation => General => Disable administration security* should be **disabled**.
o *Admin => Forum => Posts and Topics => Topic Settings => Number of posts per page in a topic page* should be configured to between 15 and 25 via  (make a note of the current value).
o *Admin => Forum => Attachments and Avatars => Attachment Settings* - Maximum size per attachment, Reformat non-JPEG images to JPEG, JPEG quality factor, Maximum width of attached images and Maximum height of attached images should be configured.
o *Admin => Forum => Attachments and Avatars => File Maintenance => Attachment Integrity Check* should be run and issues corrected.

**Important notes for batch resizing:**

- Non-JPEG images will **only** be reformatted to JPEG if the *Reformat non-JPEG images to JPEG* option is enabled.
- Existing *'attachments'* database table will be copied to *'attachmentsPreREI'* database table. *
- Original image files are saved to the *'attachmentsPreREI'* directory/folder. *
- File timestamp for resized image file is set to match timestamp of original image file.
- Batch processing progress information is displayed and updated.
- The results of the batch resizing process will be displayed on completion and also written to the forum error log file.
- The *'attachmentsPreREI'* database table and directory/folder can be removed after checks have been done to confirm successful resizing - this will free up additional disk space on the server.

The forum should be configured as follows **after** batch resizing has finished:

- *Admin => Configuration => Security and Moderation => General => Disable administration security* should be **enabled**.
- *Admin => Configuration => Server Settings => General => Enable Maintenance Mode* should be **disabled**.
- *Admin => Forum => Posts and Topics => Topic Settings => Number of posts per page in a topic page* should be configured to previous value.

## Admin Changes
In **Admin => Forum => Attachments and Avatars**:

- **Browse Files**: There is a new column (and associated button) for rotating/flipping images - the options for rotating/flipping images are only shown for image attachments.  There is also a new horizontal tab labelled *'Resize Existing Images'* that can be used for manually resizing/reformatting existing images.
- **Attachment Settings**: Options to enable/disable automatic image rotation and automatic resizing of existing images, option to reformat non-JPEG images to JPEG and options to set JPEG quality factor and max width/height values for attached images.
- **File Maintenance**: There is a new section for *Batch Resize Existing Images*.  There is also a new section for *Reset Orientation Flag*.

## Compatibility Notes
This mod was tested on SMF 2.0.15 but should work on SMF 2.0 and up.
It has also been tested on SMF 2.1 Beta 3 and there are some known issues with manual image rotation not always working and with the manual image rotation options not being displayed in posts.
This mod is not currently compatible with SMF 2.1 Beta 4.
SMF 1.x is not and will not be supported.

The [Image Processing Memory Limit](https://custom.simplemachines.org/mods/index.php?mod=4111) mod should be installed if 'white screen' issues are encountered when uploading and/or rotating images.

The [Improved Attachment Error Handling](https://custom.simplemachines.org/mods/index.php?mod=3255) mod (if so desired) should be installed **BEFORE** this mod to avoid install errors.

The [Resize Attachment Images](https://custom.simplemachines.org/mods/index.php?mod=2206) mod **MUST** be uninstalled prior to installing this version as the two mods are not compatible (this mod contains similar and updated functionality).

## Translators

- Dutch: [@rjen](https://www.simplemachines.org/community/index.php?action=profile;u=287786)
- Spanish Latin: [Rock Lee](https://www.simplemachines.org/community/index.php?action=profile;u=322597).

## Special Credit
This mod relies on the [phpExifRW](http://www.phpclasses.org/package/1042-PHP-EXIF-information-reader-and-writer.html) class, which is licensed under the [GNU Lesser General Public License](http://www.gnu.org/licenses/old-licenses/lgpl-2.1.en.html), in order to read the EXIF information from image files.  This class makes the requirement of having EXIF support built-in, which some servers do not have, not important to the task of successfully pulling the orientation out of the image file.  The **exifReader.inc** file was renamed to **Class-exifReader.php** in order to name the file in accordance with the naming convention of SMF and included in this mod.

Test images with EXIF orientation values embedded in them are available at [Galloway.me.uk](http://www.galloway.me.uk/2012/01/uiimageorientation-exif-orientation-sample-images/) and at the [ Image Orientation](http://www.elkarte.net/community/index.php?topic=2509.0) thread over at the ElkArte forum.

## Changelog
The changelog can be seen at [XPtsp.com](http://www.xptsp.com/board/free-modifications/automatic-attachment-rotation/msg9/#msg9).

## License
Copyright (c) 2016 - 2018, Douglas Orend
All rights reserved.


Changelog.txt
v5.7 - October 10th, 2018
o No functionality change.
o Updated documentation to point to new website.

v5.6 - May 21st, 2018
o Fixed undefined variable error in [b]Class-exifReader.php[/b].

V5.5 - May 7th, 2018
o Per [url=https://www.simplemachines.org/community/index.php?topic=544111.msg3970888#msg3970888]Rock Lee[/url] - updated/added Spanish langage files.
o Per [url=https://www.simplemachines.org/community/index.php?topic=544111.msg3971411#msg3971411]The QE2 Story Forum[/url] and [url=https://www.simplemachines.org/community/index.php?topic=544111.msg3971413#msg3971413]Tonyvic[/url] - added note about [i]"Attachment Settings => Re-encode potentially dangerous image attachments"[/i] setting in relation to this mod.
o Resolved issue where correct JPEG Quality value was not always being stored in database.

V5.4 - May 1st, 2018
o Per [url=https://www.simplemachines.org/community/index.php?topic=544111.msg3970720#msg3970720]@rjen[/url] - updated Dutch language files.
o Per [url=https://www.simplemachines.org/community/index.php?topic=544111.msg3970759#msg3970759]Rock Lee[/url] - fixed text formatting error for [i]attachmentsPreREI[/i] directory name text string in language files and updated help text/description relating to the saving of original image files to the [i]attachmentsPreREI[/i] directory.

V5.3 - April 30th, 2018
o Per [url=https://www.simplemachines.org/community/index.php?topic=544111.msg3970490#msg3970490]@rjen[/url] - updated Dutch language files for 'Resize Exisitng Images'.
o Updated English language files for 'Resize Exisitng Images' to provide more detailed description for manual/batch options.
o Changed behaviour of automatic image resizing when displaying a post so that images are [b]only[/b] resized if [i]'Resize existing images'[/i] is selected (in previous versions of this mod they were also resized when [i]'Reformat non-JPEG images to JPEG'[/i] was selected).
o Resolved issue that was causing image files saved with '_JPG' in their filename to be incorrectly renamed to '_jpg' in the database (this only affected image attachments created with SMF 1.x).
o Added option to keep backup of original image file when Attachment Setting [i]'Resize exisitng images'[/i] is enabled.
o Updated image resizing code for automatic, manual and batch resizing modes to create backup of the original image file and restore it if the new image file is empty.

V5.2 - April 28th, 2018
o Per [url=https://www.simplemachines.org/community/index.php?topic=544111.msg3970435#msg3970435]@rjen[/url] - fixed image and post links for images in Browse files > Resize Existing Images list.
o Per [url=https://www.simplemachines.org/community/index.php?topic=544111.msg3970435#msg3970435]@rjen[/url] - added Dutch language files.

V5.1 - April 28th, 2018
o Per [url=https://www.simplemachines.org/community/index.php?topic=525705.msg3970385#msg3970385]@rjen[/url] - fixed issue that prevented [url=https://custom.simplemachines.org/mods/index.php?mod=3770]Post and PM Inline Attachment[/url] inline attachment numbers from being incremented when modifying a post.
o Per [url=https://www.simplemachines.org/community/index.php?topic=544111.msg3970388#msg3970388]@rjen[/url] - fixed issues with sorting and browsing list of images to be resized/reformatted.
o Per [url=https://www.simplemachines.org/community/index.php?topic=544111.msg3970392#msg3970392]@rjen[/url] - load required language file rather than add text strings to other language files.

v5.0 - April 27th, 2018
o Added option for resizing existing images in either manual or batch mode.
o Changed code to only display rotate/flip options for image attachments when modifying a post.
o Fixed problem with readme file that was causing it not to display during package installation.

v4.7 - April 12th, 2018
o Updated Dutch translation, courtsey of [url=https://www.simplemachines.org/community/index.php?action=profile;u=287786]@rjen[/url]
o Per [url=https://www.simplemachines.org/community/index.php?topic=544111.msg3968843#msg3968843]@rjen[/url] fixed issue where attachment is deleted if checkbox is selected and [i]Rotate/Flip Images[/i] button is clicked.
o Per [url=https://www.simplemachines.org/community/index.php?topic=544111.msg3968843#msg3968843]@rjen[/url] included English UTF8 language file.

Subs-AutoRotation.php
<?php
/**********************************************************************************
* Subs-AutoRotation.php														   *
***********************************************************************************
* This mod is licensed under the 2-clause BSD License, which can be found here:   *
*	http://opensource.org/licenses/BSD-2-Clause								   *
***********************************************************************************
* This program is distributed in the hope that it is and will be useful, but	  *
* WITHOUT ANY WARRANTIES; without even any implied warranty of MERCHANTABILITY	  *
* or FITNESS FOR A PARTICULAR PURPOSE.											  *
**********************************************************************************/
if (!defined('SMF'))
	die('Hacking attempt...');

if (!function_exists('imageflip'))
{
	define('IMG_FLIP_HORIZONTAL', 0);
	define('IMG_FLIP_VERTICAL', 1);
	define('IMG_FLIP_BOTH', 2);
}

//==============================================================================
// Function that gets the orientation stated in the image:
//==============================================================================
function AutoRotation_GetOrientation($filename)
{
	global $sourcedir;
	if (function_exists('exif_read_data'))
	{
		// FASTER: Find this with native code if native function already exists...
		$exif = @exif_read_data($filename);
		if (isset($exif['IFD0']['Orientation']))
			return $exif['IFD0']['Orientation'];
		elseif (isset($exif['Orientation']))
			return $exif['Orientation'];
	}
	else
	{
		// SLOWER: Find this with the class if native function doesn't exist...
		require_once($sourcedir . '/Class-exifReader.php');
		$er = new phpExifReader($filename);
		$er->processFile();
		return isset($er->ImageInfo[TAG_ORIENTATION]) ? $er->ImageInfo[TAG_ORIENTATION] : 0;
	}
}

//==============================================================================
// Function dealing with Auto-Rotation of attachments:
//==============================================================================
function AutoRotation_Process($filename, $orientation = false, $jpegQuality = 100)
{
	global $context, $modSettings;
	static $default_formats = array(
		'1' => 'gif',
		'2' => 'jpeg',
		'3' => 'png',
		'6' => 'bmp',
		'15' => 'wbmp'
	);

	// Does the file even exist?  If not, return false!!
	if (!file_exists($filename))
		return false;

	// A known and supported format?
	$info = @getimagesize($filename);
	if (!$info)
		return false;
	$width = $info[0];
	$height = $info[1];
	$format = isset($default_formats[$info[2]]) ? $default_formats[$info[2]] : false;
	if (!$format || !function_exists('imagecreatefrom' . $format))
		return false;

	// Make sure it is an orientation that we can fix:
	if ($orientation === false)
		$orientation = AutoRotation_GetOrientation($filename);
	if ($orientation < 2 || $orientation > 8)
		return 1;

	// Let's attempt to allocate enough memory for this:
	if (!empty($modSettings['ipmemlimit']) && $modSettings['ipmemlimit'] > 0)
	{
		// Image Processing Memory Limit
		// Do we know the bit depth?
		if (!empty($info['bits']) && !empty($info['channels']))
			$bitDepth = $info['bits'] * $info['channels'];
		else
			$bitDepth = 96; // Probally a bad idea assuming, was 48 now 96 for 32bpp.

		// How much do we need and how much have we used. 8 bits to bytes, 65536 64k and 1.8 to add a little extra.
		$need = round(((($info[0] * $info[1] * $bitDepth / 8) + 65536 + memory_get_usage()) * 1.8) / 1048576);
		$have = (int) ini_get('memory_limit');

		// We need to set a hard limit, can't have them using all the memory
		if ($need > $modSettings['ipmemlimit'])
			return false;

		// Lets not care how much we have and just set it.
		if (ini_set('memory_limit', $modSettings['ipmemlimit'].'M') === false && $need > $have)
			return false; // Server rejected
	}
	elseif (function_exists('memory_get_usage'))
	{
		// Memory calculation goes here:
		$info['bits'] = !empty($info['bits']) ? $info['bits'] : 8;
		$info['channels'] = !empty($info['channels']) ? $info['channels'] : 4;
		$memoryNeeded = ceil($width * $height * $info['bits'] * $info['channels'] / 8 * 2);
		$currentMemory = memory_get_usage() + $memoryNeeded;
		$memoryLimit = (int) ((float) ini_get('memory_limit'));
		$memoryRequested = $memoryLimit + ceil(($currentMemory - $memoryLimit * pow(1024, 2)) / pow(1024, 2));
		
		// Do we have enough memory?  If not, try to allocate enough:
		if ($currentMemory > (integer) ini_get('memory_limit') * pow(1024, 2))
			@ini_set('memory_limit', $memoryRequested);

		// Were we successful?  If not, log the error if option is set:
		if ($currentMemory > (integer) ini_get('memory_limit') * pow(1024, 2))
		{
			if (!empty($modSettings['AutoRotation_log_error']))
			{
				loadLanguage('AutoRotation');
				log_error(sprintf('AutoRotation_memory_issue', $memoryRequested));
			}
			return false;
		}
	}
	else
	// FALLBACK METHOD: Try to allocate 256MB of memory....
		@ini_set('memory_limit', '256M');

	// Load up the image.  Abort if failure:
	$imagecreatefrom = 'imagecreatefrom' . $format;
	if (!($src_img = @$imagecreatefrom($filename)))
		return false;

	// Rotate and/or flip the image so that it is correct:
	$success = true;
	if ($orientation == 2)		// -> Flipped horizontally
		$success = imageflip($src_img, IMG_FLIP_HORIZONTAL);
	elseif ($orientation == 3)	// -> Rotated 180 degrees clockwise
		$src_img = imagerotate($src_img, 180, 0);
	elseif ($orientation == 4)	// -> Flipped vertically
		$success = imageflip($src_img, IMG_FLIP_VERTICAL);
	elseif ($orientation == 5)	// -> Flipped vertically, then rotated 90 degrees counterclockwise
	{
		if ($success = imageflip($src_img, IMG_FLIP_VERTICAL))
			$src_img = imagerotate($src_img, -90, 0);
	}
	elseif ($orientation == 6)	// -> Rotated 90 degrees counterclockwise
		$src_img = imagerotate($src_img, -90, 0);
	elseif ($orientation == 7)	// -> Flipped horizontally, then rotated 90 degrees counterclockwise
	{
		if ($success = imageflip($src_img, IMG_FLIP_HORIZONTAL))
			$src_img = imagerotate($src_img, -90, 0);
	}
	elseif ($orientation == 8)	// -> Rotated 90 degrees clockwise
		$src_img = imagerotate($src_img, 90, 0);

	// Make sure that we are still dealing with an image in memory!
	if ($success && is_resource($src_img))
	{
		if (function_exists($func = 'image' . $format))
			$success = @$func($src_img, $filename);
		else
			$success = imagejpeg($src_img, $filename, $jpegQuality);
	}
	if (is_resource($src_img))
		imagedestroy($src_img);
	return ($success ? $orientation : false);
}

//==============================================================================
// Function dealing with Auto-Rotation of attachments during downloads:
//==============================================================================
function AutoRotation_Download($img_name, $id_thumb, $img_type)
{
	global $smcFunc, $context;

	// Rotate and/or flip the image according to EXIF information:
	$orientation = -1;
	if ($img_type == 3)
	{
		// This is a thumbnail!  Drats...  Gotta find the original and process it, too.... :(
		$request = $smcFunc['db_query']('', '
			SELECT a.id_folder, a.filename, a.file_hash, a.id_attach
			FROM {db_prefix}attachments AS a
				INNER JOIN {db_prefix}attachments AS thumb ON (a.id_thumb = thumb.id_attach)
			WHERE thumb.id_attach = {int:thumbnail}
			LIMIT 1',
			array(
				'thumbnail' => $id_thumb,
			)
		);
		while ($row = $smcFunc['db_fetch_assoc']($request))
		{
			// Prevent other instances of the forum from processing the image right now:
			AutoRotation_Update($id_attach);

			// Let's process this full image properly now, then update the database:
			$attachment = getAttachmentFilename($row['filename'], $row['id_attach'], $row['id_folder'], false, $row['file_hash']);
			$orientation = AutoRotation_Process($attachment);
			$size = @getimagesize($attachment);
			AutoRotation_Update($id_attach, $size[0], $size[1], $orientation);
		}
		$smcFunc['db_free_result']($request);
	}

	// Prevent other instances of the forum from processing the image right now:
	AutoRotation_Update($id_thumb);

	// Let's process this thumbnail properly now, then update the database:
	AutoRotation_Process($img_name, $orientation);
	$size = @getimagesize($img_name);
	AutoRotation_Update($id_thumb, $size[0], $size[1], $orientation);
}

//==============================================================================
// Support functions:
//==============================================================================
function AutoRotation_Update($id_attach, $width = 0, $height = 0, $orientation = 1, $jpegQuality = 100)
{
	global $smcFunc;
	$smcFunc['db_query']('', '
		UPDATE {db_prefix}attachments
		SET proper_rotation = {int:rotated}' .
			(!empty($width) ? ', width = {int:width}' : '') .
			(!empty($height) ? ', height = {int:height}' : '') .
			(!empty($jpegQuality) ? ', jpeg_quality = {int:jpeg_quality}' : '') . '
		WHERE id_attach = {int:id_attach}',
		array(
			'id_attach' => (int) $id_attach,
			'rotated' => max(1, (int) $orientation),
			'width' => (int) $width,
			'height' => (int) $height,
			'jpeg_quality' => (int) $jpegQuality
		)
	);
}

//==============================================================================
// Function dealing with Auto-Rotation of attachments during display:
//==============================================================================
function AutoRotation_Display($row)
{
	global $smcFunc, $modSettings;

	// Is this an image OR has already been processed?  Abort if not:
	if (empty($row['width']) || empty($row['height']) || !empty($row['img_rotation']))
		return $row;

	// Rotate this image if necessary:
	$img = getAttachmentFilename($row['filename'], $row['id_attach'], $row['id_folder'], false, $row['file_hash']);
	$orientation = AutoRotation_Process($img);
	$size = @getimagesize($img);
	$jpegQuality = min($row['jpeg_quality'], empty($modSettings['attachment_jpeg_quality']) ? 100 : $modSettings['attachment_jpeg_quality']);
	AutoRotation_Update($row['id_attach'], $row['width'] = $size[0], $row['height'] = $size[1], $orientation, $jpegQuality);

	// If there is a thumbnail that needs rotating, then let's rotate the thumbnail:
	if (!empty($row['thumb_id']) && $orientation > 1)
	{
		// Prevent other instances of the forum from processing the image right now:
		AutoRotation_Update($row['thumb_id']);

		// Let's process this thumbnail properly now, then update the database:
		$img = getAttachmentFilename($row['thumb_filename'], $row['thumb_id'], $row['attach_folder'], false, $row['thumb_hash']);
		AutoRotation_Process($img, $orientation);
		$size = @getimagesize($img);
		AutoRotation_Update($row['thumb_id'], $row['thumb_width'] = $size[0], $row['thumb_height'] = $size[1], $orientation);
	}

	// Return all this information back to the caller:
	return $row;
}

//==============================================================================
// This replacement "imageflip" function gets used when using PHP < 5.5.
//==============================================================================
// Credit: Daniel Klein => http://php.net/manual/en/function.imageflip.php#118774
//==============================================================================
if (!function_exists('imageflip'))
{
	function imageflip($image, $mode)
	{
		if ($mode == IMG_FLIP_HORIZONTAL)
		{
			$max_x = imagesx($image) - 1;
			$half_x = $max_x / 2;
			$sy = imagesy($image);
			$temp_image = imageistruecolor($image) ? imagecreatetruecolor(1, $sy) : imagecreate(1, $sy);
			for ($x = 0; $x < $half_x; ++$x)
			{
				imagecopy($temp_image, $image, 0, 0, $x, 0, 1, $sy);
				imagecopy($image, $image, $x, 0, $max_x - $x, 0, 1, $sy);
				imagecopy($image, $temp_image, $max_x - $x, 0, 0, 0, 1, $sy);
			}
		}
		elseif ($mode == IMG_FLIP_VERTICAL)
		{
			$sx = imagesx($image);
			$max_y = imagesy($image) - 1;
			$half_y = $max_y / 2;
			$temp_image = imageistruecolor($image) ? imagecreatetruecolor($sx, 1) : imagecreate($sx, 1);
			for ($y = 0; $y < $half_y; ++$y)
			{
				imagecopy($temp_image, $image, 0, 0, 0, $y, $sx, 1);
				imagecopy($image, $image, 0, $y, 0, $max_y - $y, $sx, 1);
				imagecopy($image, $temp_image, 0, $max_y - $y, 0, 0, $sx, 1);
			}
		}
		elseif ($mode == IMG_FLIP_BOTH)
		{
			$sx = imagesx($image);
			$sy = imagesy($image);
			$temp_image = imagerotate($image, 180, 0);
			imagecopy($image, $temp_image, 0, 0, 0, 0, $sx, $sy);
		}
		else
			return false;

		imagedestroy($temp_image);
		return true;
	}
}

//==============================================================================
// This replacement "imagerotate" function gets used when using PHP < 4.3.
// NOTE: Parameters "bgd_color" and "ignore_transparent" are completely ignored!
//==============================================================================
// Credit: Ajenbo => http://php.net/manual/en/function.imagerotate.php#85329
//==============================================================================
if (!function_exists('imagerotate'))
{
	function imagerotate($src_img, $angle, $bgd_color = 0, $ignore_transparent = 0)
	{
		$src_x = imagesx($src_img);
		$src_y = imagesy($src_img);
		if ($angle == 180)
		{
			$dest_x = $src_x;
			$dest_y = $src_y;
		}
		elseif ($src_x <= $src_y)
		{
			$dest_x = $src_y;
			$dest_y = $src_x;
		}
		elseif ($src_x >= $src_y)
		{
			$dest_x = $src_y;
			$dest_y = $src_x;
		}

		$rotate = imagecreatetruecolor($dest_x, $dest_y);
		imagealphablending($rotate, false);

		switch ($angle)
		{
			case 270:
				for ($y = 0; $y < $src_y; $y++)
				{
					for ($x = 0; $x < $src_x; $x++)
					{
						$color = imagecolorat($src_img, $x, $y);
						imagesetpixel($rotate, $dest_x - $y - 1, $x, $color);
					}
				}
				break;
			case 90:
				for ($y = 0; $y < $src_y; $y++)
				{
					for ($x = 0; $x < $src_x; $x++)
					{
						$color = imagecolorat($src_img, $x, $y);
						imagesetpixel($rotate, $y, $dest_y - $x - 1, $color);
					}
				}
				break;
			case 180:
				for ($y = 0; $y < $src_y; $y++)
				{
					for ($x = 0; $x < $src_x; $x++)
					{
						$color = imagecolorat($src_img, $x, $y);
						imagesetpixel($rotate, $dest_x - $x - 1, $dest_y - $y - 1, $color);
					}
				}
				break;
			default: 
				$rotate = $src_img; 
		}
		return $rotate; 
	}
}

//==============================================================================
// Admin functions for dealing with image rotation
//==============================================================================
function AutoRotation_LoadHelp()
{
	loadLanguage('AutoRotation');
}

function AutoRotation_AdminHook(&$subActions)
{
	global $context;
	loadLanguage('AutoRotation');
	if (isset($subActions['remove']))
		$context['autorotation_hook'] = $subActions['remove'];
	$subActions['remove'] = 'AutoRotation_Rotate';
	$subActions['clearflags'] = 'AutoRotation_Clear';
}

function AutoRotation_Rotate()
{
	global $smcFunc, $context, $modSettings;

	// First, let's make sure that the session is valid!!!
	checkSession('post');

	// Start any user-requested image processing....
	if (!empty($_POST['orient']))
	{
		// Sanitize the inputs before we pass it to the database:
		$attach = array();
		foreach ($_POST['orient'] as $id => $orient)
			$attach[(int) $id] = (int) $orient;

		// This is a thumbnail!  Drats...  Gotta find the original and process it, too.... :(
		$request = $smcFunc['db_query']('', '
			SELECT
				a.id_attach, a.id_folder, a.file_hash, a.filename, a.jpeg_quality, t.id_attach AS thumb_id,
				t.id_folder AS thumb_folder, t.file_hash AS thumb_hash, t.filename AS thumb_name
			FROM {db_prefix}attachments AS a
				LEFT JOIN {db_prefix}attachments AS t ON (t.id_attach = a.id_thumb)
			WHERE a.id_attach IN ({array_int:attach_list})
				AND a.width > {int:zero}
				AND a.height > {int:zero}',
			array(
				'attach_list' => array_keys($attach),
				'zero' => 0,
			)
		);
		while ($row = $smcFunc['db_fetch_assoc']($request))
		{
			// Prevent other instances of the forum from processing the image right now:
			AutoRotation_Update($row['id_attach']);

			// Let's process this image properly now, then update the database:
			$img = getAttachmentFilename($row['filename'], $row['id_attach'], $row['id_folder'], false, $row['file_hash']);
			$jpegQuality = min($row['jpeg_quality'], empty($modSettings['attachment_jpeg_quality']) ? 100 : $modSettings['attachment_jpeg_quality']);
			AutoRotation_Process($img, $orientation = $attach[$row['id_attach']], $jpegQuality);
			$size = @getimagesize($img);
			AutoRotation_Update($row['id_attach'], $size[0], $size[1], $orientation, $jpegQuality);

			// Also process the thumbnail (if available):
			if (!empty($row['thumb_id']))
			{
				// Prevent other instances of the forum from processing the image right now:
				AutoRotation_Update($row['thumb_id']);

				// Let's process this thumbnail properly now, then update the database:
				$img = getAttachmentFilename($row['thumb_name'], $row['thumb_id'], $row['thumb_folder'], false, $row['thumb_hash']);
				AutoRotation_Process($img, $orientation);
				$size = @getimagesize($img);
				AutoRotation_Update($row['thumb_id'], $size[0], $size[1], $orientation);
			}
		}
		$smcFunc['db_free_result']($request);
	}

	// Since we're chain-calling stuff, this MUST be called after processing the images!!!
	if (isset($context['autorotation_hook']) && is_callable($context['autorotation_hook']))
		$context['autorotation_hook']();
}

function AutoRotation_Clear()
{
	global $smcFunc, $context;

	// Make sure that the session is valid and the user can manage attachments:
	checkSession('get');
	isAllowedTo('manage_attachments');

	// Clear orientation flags, then go back to maintenance screen:
	$smcFunc['db_query']('', '
		UPDATE {db_prefix}attachments
		SET proper_rotation = {int:rotated}',
		array(
			'rotated' => 0,
		)
	);
	redirectexit('action=admin;area=manageattachments;sa=maintenance;' . $context['session_var'] . '=' . $context['session_id']);
}

//==============================================================================
// Image Resizing functionality added by GL700Wing:
//==============================================================================
function AutoRotation_Inbound($index, $pm = false)
{
	global $modSettings, $sourcedir;

	// Does this index actually exist?  If not, return to caller:
	if (!isset($_FILES['attachment']['tmp_name'][$index]))
		return;

	// Figure some things out:
	$pre = $pm ? 'pm_' : '';
	$filename = $_FILES['attachment']['tmp_name'][$index];
	$sizeLimit = empty($modSettings[($pm ? 'pmA' : 'a') . 'ttachmentSizeLimit']) ? 0 : max(0, $modSettings[($pm ? 'pmA' : 'a') . 'ttachmentSizeLimit']) * 1024;
	$jpegQuality = min(100, empty($modSettings[$pre . 'attachment_jpeg_quality']) ? 100 : $modSettings[$pre . 'attachment_jpeg_quality']);
	$imageRotated = false;

	// Find out image mime.
	$imageSize = @getimagesize($filename);

	// If it's a JPEG image rotate it if necessary.
	if (!empty($modSettings['attachment_auto_rotate']) && $imageSize['mime'] == 'image/jpeg')
	{
		// Disable image reencoding to enable image rotation.
		$modSettings['attachment_image_reencode'] = false;

		@ini_set('gd.jpeg_ignore_warning', 1);
		$image = @imagecreatefromjpeg($filename);

		// Rotate JPEG image according to Exif orientation.
		$imageExif = @exif_read_data($filename);
		if (isset($imageExif['Orientation']) || array_key_exists('Orientation', $imageExif))
		{
			if ($_FILES['attachment']['size'][$index] <= $sizeLimit)
			{
				$imageOrientation = $imageExif['Orientation'];
				switch($imageOrientation)
				{
					case 3:
						$image = imagerotate($image, 180, 0);
						break;
					case 6:
						$image = imagerotate($image, -90, 0);
						break;
					case 8:
						$image = imagerotate($image, 90, 0);
						break;
				}

				$imageRotated = true;
			}
		}

		// Create the JPEG file with maximum quality of 100%.
		imagejpeg($image, $filename, $jpegQuality);
		imagedestroy($image);
	}

	// Resize/reformat image and rotate it if necessary.
	if (list ($width, $height, $type) = @getimagesize($filename))
	{
		if (AutoRotation_Aspect($width, $height, $pre) || empty($sizeLimit) || $jpegQuality != 100)
		{
			// If the image should be reformatted to JPEG.
			// Note: By default BMP images are reformatted to JPEG by resizeImageFile().
			if (!empty($modSettings[$pre . 'attachment_image_reformat']) || $type == 6 )
				$preferred_format = 2;
			else
				$preferred_format = $type;

			// Resize the image (and rotate it if necessary).
			require_once($sourcedir . '/Subs-Graphics.php');
			if (resizeImageFile($filename, $filename . '.temp', $width, $height, $preferred_format, $jpegQuality))
			{
				// Delete the old tmp file and rename the new one.
				unlink($filename);
				rename($filename . '.temp', $filename);

				// If it's a JPEG image rotate and/or compress the image.
				if (!empty($modSettings['attachment_auto_rotate'])  && ($imageSize['mime'] == 'image/jpeg' && !$imageRotated))
				{
					// Disable image reencoding to enable image rotation.
					$modSettings['attachment_image_reencode'] = false;

					@ini_set ('gd.jpeg_ignore_warning', 1);
					$image = @imagecreatefromjpeg($filename);

					// Rotate JPEG image according to Exif orientation.
					if (isset($imageExif['Orientation']) || array_key_exists('Orientation', $imageExif))
					{
						$imageOrientation = $imageExif['Orientation'];
						switch($imageOrientation)
						{
							case 3:
								$image = imagerotate($image, 180, 0);
								break;
							case 6:
								$image = imagerotate($image, -90, 0);
								break;
							case 8:
								$image = imagerotate($image, 90, 0);
								break;
						}
					}

					// Create the JPEG file with maximum quality of 100%.
					imagejpeg($image, $filename, $jpegQuality);
					imagedestroy($image);
				}

				// Recheck the file size
				$_FILES['attachment']['size'][$index] = filesize($filename);

				// Change the file suffix to 'jpg' if necessary.
				$filename = $_FILES['attachment']['name'][$index];
				if ($preferred_format == 2 && strrchr($filename, '.') != '.jpg')
					$_FILES['attachment']['name'][$index] = substr($filename, 0, -(strlen(strrchr($filename, '.')))) . '.jpg';
			}
		}
	}
}

function AutoRotation_Aspect(&$width, &$height, $pre = '')
{
	global $modSettings;

	// If image dimensions are equal to or below admin settings, don't continue:
	$max_width = !empty($modSettings[$pre . 'attachment_image_width']) ? $modSettings[$pre . 'attachment_image_width'] : $width;
	$max_height = !empty($modSettings[$pre . 'attachment_image_height']) ? $modSettings[$pre . 'attachment_image_height'] : $height;
	if ($width <= $max_width && $height <= $max_height)
		return false;

	// Adjust size of image, maintaining aspect ratio of image.  Note that the code rounds
	// the decimals up to the next number, to avoid creating images with a width/height of 0!
	$aspect_ratio = min($max_width / $width, $max_height / $height);
	$width  = (int) ceil($width * $aspect_ratio);
	$height = (int) ceil($height * $aspect_ratio);
	return true;
}

?>
